### PR TITLE
Transpose scheduler vectorization width patch

### PR DESCRIPTION
**Pull request by jjsjann123**
This PR sits on top of #753 

1. refactored transpose scheduler to compute vectorzation width using aggregated inner dimensions, rather than only considering the inner most dimensions.

This is best explained with an example:
```
tv2 (i0, i1, i2, i3, i4)
tv1 (i3, i4, i2, i0, i1) = tv2.permute(3, 4, 2, 0, 1)
```
Transpose scheduler utilizes tile on the fastest dimension of tv1 (i1) and tv2 (i2).

When small transpose dimensions are present, we map the two iter domain on tv1 and merge each with their neighboring iter domains. (we also would split iter domain, when merged results is unbalanced, but we are not discussing or handling that here yet. See TODOs below).
So in the example above, if we are to have merged (i3*i4), and (i0*i1), when computing vectorization width, we need to map the merged iter domains on each tensor and use that to compute vectorization width.
More details can be referred to in Note: [Computing Vectorization Width for Transpose]

2. updated mergeDims logic to merge dimensions in specified order.
Previously we sort dimension indices before merging, this violates the virtual inner most dimension analysis, which relies on the order of merge for virtual inner most dimension.
Fixes: #675

TODOs in future PR:
- update iter domain mapping, so we can consider split dimensions for vectorization;
- for group2 output, we don't need a global vectorization width, but could rather have each tensor view keep a unique vectorization instead;

### CODE REVIEW

File: csrc/scheduler/transpose.cpp [hunk 1]
```diff
@@ -714,29 +714,138 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
+  // Currently there's limitation on our iter domain mapping. Since we can only
+  // do it on rfactor/root domain, we cannot map across `split` domains. So the
+  // example above will only have vectorization size of 2 and 4 repsectively for
+  // the merge virtual innermost dimensions, rather than considering the split
+  // and merged i2/2 & 2.
+  //
+  // TODO: We use ContiguousInnerDimensionsMapper to compute the size of virtual
+  // innermost dimension. The analysis right now is limited on rfactor domain
+  // only, so we can't actually map the `split` iter domains, which limits the
+  // vectorization width we can apply. We need to fix that.
```
Comment (zasdfgbnm) [inline csrc/scheduler/transpose.cpp:740]: I think the problem is bigger than that. Because the vectorization factor can be incompatible with tiling. For example, what if the tile size is `7`? (Could you add a test for it? Also, it worth testing the case where the virtual inner dim is 1024*7 but the tile size is 2*7). I think the TODO item for starting propagating from leaf instead of rFactor domain is required, not optional or future work, because otherwise we can not correctly handle tiling? Also, it would make the analysis here much easier: we can factor out part of `scheduleTranspose` out, and just apply it here on `cloned_1_td` and begin the analysis on leaf. If so, probably we should do it in a separate PR again, and test it separately?  There are a lot of vectorization helper tests in https://github.com/NVIDIA/Fuser/blob/main/test/test_gpu_utils.cpp

File: csrc/scheduler/utils.h [hunk 2]
```diff
@@ -40,6 +40,14 @@ constexpr int64_t y_grid_limit = 65535;
 constexpr int64_t z_grid_limit = 65535;
 constexpr int64_t z_block_limit = 64;
 
+constexpr int64_t maxVectorizationWidth(int64_t n) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/scheduler/utils.h:43]: this is not `lastPow2` since this guy requires `n % next_vector_size == 0` at each step check.

File: csrc/scheduler/transpose.cpp [hunk 3]
```diff
@@ -714,29 +714,138 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
+          : runtime_info.expressionEvaluator().evaluate(inner_size_it->second);
+      // TODO: Do not assert here. we can just reduce vectorization size to 1 if
+      // we can't infer an inner size.
+      TORCH_INTERNAL_ASSERT(
+          tv_vectorize_factor_opt.hasValue(),
+          "Vectorization heuristic could not evaluate inner most size.");
+      int64_t tv_vectorize_factor = tv_vectorize_factor_opt.as<int64_t>();
+      vectorize_factor1 = std::min(
+          vectorize_factor1,
+          scheduler_utils::maxVectorizationWidth(tv_vectorize_factor));
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:799]: sounds good. I'll have it cleaned up.

File: csrc/scheduler/utils.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/scheduler/utils.h:43]: Ahh, true. But still `getVectorizationFactor` is replicating the same logic. I think we should make `getVectorizationFactor` use `maxVectorizationWidth`.

File: csrc/scheduler/transpose.cpp (hunk 1 above)
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:740]: I'll try to add a test for odd tile sizes. I'm not quite sure I follow how `odd` tile size would break it here. Doesn't an odd tile size mean any inner dimension mapping would be `odd` and we wouldn't have any vectorization logic enabled right now?

Other than that, all parts regarding refactor and simplify `scheduleTranspose` sounds good to me here.

**Comment by jjsjann123** [comment]
Oops. look like `NVFuserTest.FusionSegmentReduceSoftmax_CUDA` is broken by my changes... :fearful: 
A bit surprising, since I don't think I have changed any existing behavior. I'll double check when test run is done.

File: csrc/scheduler/vectorize_helper.cpp [hunk 4]
```diff
@@ -64,63 +64,96 @@ ContiguousInnerDimensionsMapper::ContiguousInnerDimensionsMapper(
...
-  for (auto id : reference->getMaybeRFactorDomain()) {
-    // Exclude reduction IDs if the reference is a fusion input as they
-    // don't manifest at all in the fusion. This simplifies the
-    // analysis in getContigMergeOfInnerSize, which only looks at
-    // non-reduction rfactor domains. Including reduction domains here
-    // can result in incorrect ordering
-    if (reference->isFusionInput() && id->isReduction()) {
-      continue;
+
+  auto projected_rfactor = projectId(ids, reference->getMaybeRFactorDomain());
```
Comment (zasdfgbnm) [inline csrc/scheduler/vectorize_helper.cpp:127]: Should we `addProjectedExtent` for `ids` before projecting to rfactor?

Comment (jjsjann123) [reply csrc/scheduler/vectorize_helper.cpp:127]: I think this isn't what the old behavior required. having said that, sounds like a good idea to add some extra bits there?!

I'm actually seeing another failure related to no extend. `RNGTest.BroadcastingRNGSmem`. Let me try this and also take another look at that failing test.

**Comment by jjsjann123** [comment]
Yay~~ all tests passed for me. I'll clean up and split the PR per Xiang's request.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
benchmark seems to have timed_out... :confused: 

Does this mean we have a regression or are we just living on the edge of time-out?!

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
I think there's something wrong with the transpose benchmark. Looks like the number of elements isn't right when running a series of benchmarks. Which was why the CI failure is not repro'ed when running individual test.

I'll fix that in a follow up PR (as well as renaming the benchmarks)

**Comment by jjsjann123** [comment]
!build

File: csrc/scheduler/transpose.cpp [hunk 5]
```diff
@@ -714,29 +718,124 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
+  // vectorization impact. i.e. when split_before_tiling, we should try to split
+  // on a factor that allows vectorization.
+  {
+    // Compute maximum vectorize factor that can be used
+    int64_t vectorize_factor1 = max_unroll_factor;
+    int64_t vectorize_factor2 = max_unroll_factor;
+
+    // duplicating reference1's TensorDomain, since the transformations applied
+    // is not persistent and only needed for us to compute vectorization width.
+    TensorDomain* cloned_1_td = IrBuilder::create<TensorDomain>(
```
Comment (zasdfgbnm) [inline csrc/scheduler/transpose.cpp:755]: Could you add a new constructor `TensorDomain::TensorDomain(TensorDomain*)` and use that ctor here? Preferably in a separate PR.

File: csrc/scheduler/transpose.cpp [hunk 6]
```diff
@@ -679,19 +679,23 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
-  int64_t max_input_dtype_size = 1;
-
-  size_t n_input_tensors = 0;
-  for (auto inp : ir_utils::filterByType<TensorView>(fusion->inputs())) {
-    max_input_dtype_size = std::max(
-        max_input_dtype_size,
-        (int64_t)dataTypeSize(inp->getDataType().value(), index_type));
-    n_input_tensors++;
-  }
+  int64_t max_io_dtype_size = 1;
```
Comment (zasdfgbnm) [inline csrc/scheduler/transpose.cpp:682]: This looks like an independent fix. Could you do it in a separate PR so that we can quickly merge it? I want to reduce the size of this PR as much as possible.

File: csrc/scheduler/transpose.cpp [hunk 7]
```diff
@@ -714,29 +718,124 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
...
+  // transpose scheduler would apply transformation and tile on virtual
+  // innermost dimensions. ( (i0*i1*i2/2), (2*i3*i4)/32) So we need to use the
+  // size of the virtual innermost dimensions to compute our vectorization
+  // width. In the example above, we are looking at 2*3*65536/2, 2*4*7.
+  //
+  // Currently there's limitation on our iter domain mapping. Since we can only
+  // do it on rfactor/root domain, we cannot map across `split` domains. So the
+  // example above will only have vectorization size of 2 and 4 repsectively for
+  // the merge virtual innermost dimensions, rather than considering the split
+  // and merged i2/2 & 2.
```
Comment (zasdfgbnm) [inline csrc/scheduler/transpose.cpp:739]: Is this still the case even with https://github.com/NVIDIA/Fuser/pull/753?

File: csrc/scheduler/utils.cpp [hunk 8]
```diff
@@ -170,21 +170,32 @@ std::optional<size_t> mergeDims(
   if (to_merge.size() == 1) {
     return to_merge[0];
   }
-  std::sort(to_merge.begin(), to_merge.end());
```
Comment (zasdfgbnm) [inline csrc/scheduler/utils.cpp:173]: Why this is changed?

File: csrc/scheduler/transpose.cpp (hunk 7 above)
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:739]: Yes that's still the case.

We need to further refactor the mapping thing in order to support that.

File: csrc/scheduler/utils.cpp (hunk 8 above)
Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:173]: We can't blindly sort and merge with the sorted order.

i.e. The transform is applied on reference1 and it needs to be populated to group2 for tiling.
if we have reference1 (i0, i1, i2, i3, i4) -> reference2 (i3, i4, i2, i0, i1).

Consider a split on i2 that gives us (i2o, i2i). So reference1 becomes (i0, i1, i2o, i2i, i3, i4) and dims_merged_with_2 is `i0, i2o`.
previously we have merge i0, i2o, and then merge i1 with them. Which gives us. (i0 * i2o * i1). This is not vectorizable on reference2 (i3, i4, i2, i0, i1).
So we need to ensure that we merge things in the logical order as they show up in `dims_merged_with_2`.

File: csrc/scheduler/transpose.cpp (hunk 7 above)
Comment (zasdfgbnm) [reply csrc/scheduler/transpose.cpp:739]: Isn't this automatically handled in `projectId`?

File: csrc/scheduler/utils.cpp (hunk 8 above)
Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:173]: https://github.com/NVIDIA/Fuser/issues/675#issuecomment-1671627684

File: csrc/scheduler/transpose.cpp (hunk 7 above)
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:739]: Hmmm. I wasn't totally sure about this one. I was still trying to have projectId to map ids to `rfacotr_domain` of the reference. I think previously I was struggling with `getProjectedExtent` as well as propogation. But that might have been fixed with us changing addProjectedExtent to before the projectId call.

I think you are saying it's not necessary and propagation should still be able to handle that. Let me give this another shot to see if it would just work.

File: csrc/scheduler/utils.cpp (hunk 8 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:173]: Thank you for the explanation. Could you also pull this out as a separate PR? I don't like packing fixes on multiple issues in the same PR. Also, in this case, does it make sense to make the `to_merge` argument an `unordered_set`? It would also be good to modify `FusionMergeDims_CUDA` to test this behavior change.

Comment (jjsjann123) [reply csrc/scheduler/utils.cpp:173]: Will do (regarding split out to a new PR).

Meanwhile, we do want to keep std::vector instead of std::unordered_set, since we can't dictate an iterating order with unordered_set.

File: csrc/scheduler/transpose.cpp [hunk 9]
```diff
@@ -714,29 +718,124 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
       ceilDiv((int64_t)params->tile_size1 * (int64_t)params->tile_size2, 32l);
   max_unroll_factor = std::min(max_unroll_factor, max_unroll_factor_block);
 
-  // Compute maximum vectorize factor that can be used
-  size_t vectorize_factor1 = max_unroll_factor;
-  size_t vectorize_factor2 = max_unroll_factor;
+  // Note: [Computing Vectorization Width for Transpose]
```
Comment (naoyam) [inline csrc/scheduler/transpose.cpp:721]: Can you please factor this part out as a function like `getTransposeVectorizationFactor`?

File: csrc/scheduler/transpose.cpp [hunk 10]
```diff
@@ -714,29 +714,138 @@ std::shared_ptr<TransposeParams> getTransposeHeuristics(
       ceilDiv((int64_t)params->tile_size1 * (int64_t)params->tile_size2, 32l);
   max_unroll_factor = std::min(max_unroll_factor, max_unroll_factor_block);
 
-  // Compute maximum vectorize factor that can be used
-  size_t vectorize_factor1 = max_unroll_factor;
-  size_t vectorize_factor2 = max_unroll_factor;
-
-  for (auto tv : grouped_inputs_outputs[0]) {
-    const auto tv_vectorize_factor =
-        runtime_info.getMaxVectorizableWidth(tv, false);
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:726]: Will do , I'll also remove the test `NHWCMaxVectorizableWidth_CUDA`.

File: csrc/scheduler/transpose.cpp (hunk 7 above)
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:738]: I think this is still the case here.

Since we project leaf domain to rfactor_domain before propagating them in `ContiguousInnerDimensionsMapper`, so for split leaf domains, they would simply vanish before we get the chance to actually propagate them.

This is also verified in the test added in this PR:
`TransposeSplitAggregatedVectorizationWidth_CUDA`

Do still plan to support this, but it's going to be another refactor on the mapper.

File: csrc/scheduler/vectorize_helper.h [hunk 11]
```diff
@@ -314,6 +314,14 @@ int64_t getVectorizationFactor(
     HeuristicSummary* data_cache,
     int64_t break_point);
 
+int64_t getVectorizationVectorTransposeGroup(
```
Comment (naoyam) [inline csrc/scheduler/vectorize_helper.h:317]: nit: we usually call it a factor, so `getVectorizationFactorTransposeGroup` would be more consistent

File: csrc/scheduler/vectorize_helper.cpp [hunk 12]
```diff
@@ -861,6 +861,50 @@ int64_t getVectorizationFactor(
...
+    size_t inner_most_dim,
+    const std::vector<size_t>& dims_to_merge,
+    const std::vector<TensorView*>& vec_tv,
+    int64_t max_vectorization) {
+  max_vectorization = scheduler_utils::maxVectorizationWidth(max_vectorization);
+  std::vector<IterDomain*> virtual_innermost_dim;
+  // find the virtual_innermost_dim in reference so we can later map
+  // that to individual TensorView in vec_tv.
+  for (const auto& dim : dims_to_merge) {
+    virtual_innermost_dim.insert(
```
Comment (naoyam) [inline csrc/scheduler/vectorize_helper.cpp:876]: Why push front instead of push back? Does the order matter?

Comment (jjsjann123) [reply csrc/scheduler/vectorize_helper.cpp:876]: the order matters, since in the other PR where we refactored `mergeDims`, we need to follow the logical order from inner to outer to have aligned memory access for vectorization

Comment (jjsjann123) [reply csrc/scheduler/vectorize_helper.cpp:876]: if you are worried about performance with the `insert` here, let me clean it up.

File: csrc/scheduler/vectorize_helper.cpp [hunk 13]
```diff
@@ -861,6 +861,50 @@ int64_t getVectorizationFactor(
...
+  auto contig_inner_map =
+      vectorize_helper::ContiguousInnerDimensionsMapper::map(
+          reference, virtual_innermost_dim)
+          .getTvToContigMergeOfInnerSizeMap();
+  for (auto tv : vec_tv) {
+    auto inner_size_it = contig_inner_map.find(tv);
+    auto tv_vectorize_factor_opt = inner_size_it == contig_inner_map.end()
+        ? 1
+        : runtime_info.expressionEvaluator().evaluate(inner_size_it->second);
+    // TODO: Do not assert here. we can just reduce vectorization size to 1 if
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/scheduler/vectorize_helper.cpp:894]: I don't know enough to answer that question. :man_shrugging: 

That's why I didn't change the assert yet. We can revisit this when the issue shows up.

File: test/test_gpu_transpose.cpp [hunk 14]
```diff
@@ -1127,4 +1127,156 @@ TEST_F(NVFuserTest, UnswitchPredicateIssueRepro667_CUDA) {
...
+  TORCH_CHECK(!runtime->isSegmented(), "Segmentation not expected");
+
+  auto t1 = at::native::view(t0, {1024, 2, 2, 3});
+  auto t2 = t1.transpose(1, 2);
+  auto ref = at::reshape(t2, {1024, 2, 6});
+
+  TORCH_CHECK(ref.equal(cg_outputs.at(0)));
+}
+
+// TODO: this one currently breaks when I have reshape+transpose scheduler.
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply test/test_gpu_transpose.cpp:1205]: Yes it is. It's a projection issue, I'm working on this one right now.
If failes here: https://github.com/NVIDIA/Fuser/blob/6d2c15719f222c562d0a28c3dd54620b5491c5a1/csrc/scheduler/transpose.cpp#L572-L573
```
RuntimeError: mapped_id != nullptr INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/scheduler/transpose.cpp":135,
```

It actually shows up in nanogpt backward. Kevin accidentally run into this as well when he was playing with this branch earlier.

You won't see the failure here though. Let me move these tests to the other PR, they seems to be quite distracting here.

File: csrc/scheduler/vectorize_helper.cpp (hunk 12 above)
Comment (naoyam) [reply csrc/scheduler/vectorize_helper.cpp:876]: I understand the ordering does matter for `mergeDims`, but we are not calling it with this `virtual_innermost_dim`, right?

File: csrc/scheduler/transpose.cpp (hunk 7 above)
Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:738]: Since we have this `auto projected_rfactor = projectId(filtered_ids, rfactor_domain);` before propagation.
```
T1_g[ iS5{i3}, iS6{( ceilDiv(i2, 5) )}, iS7{5}, iS3{i0} ]
 root domain : (iS3{i0}, iS4{i2}, iS5{i3})
 rfactor domain : (iS5{i3}, iS4{i2}, iS3{i0})
 contiguity: t t t
  Split: iS4{i2} by factor 5 -> iS6{( ceilDiv(i2, 5) )}, iS7{5}, start offset: 0, stop offset: 0
 leaf domain : (iS5{i3}, iS6{( ceilDiv(i2, 5) )}, iS7{5}, iS3{i0})
```

```
(gdb) print filtered_ids
$9 = std::vector of length 2, capacity 2 = {0x55555acf9050, 0x555559df9970}
(gdb) print projected_rfactor
$10 = std::vector of length 1, capacity 2 = {0x555559df9970}
(gdb) print filtered_ids[0]->toString(0)
$5 = "iS7{5}"
(gdb) print filtered_ids[1]->toString(0)
$6 = "iS3{i0}"
(gdb) print projected_rfactor[0]->toString(0)
$4 = "iS3{i0}"
```
So the split dimension `iS7{5}` is lost when we project it back.

File: csrc/scheduler/vectorize_helper.cpp (hunk 12 above)
Comment (jjsjann123) [reply csrc/scheduler/vectorize_helper.cpp:876]: Sorry I wasn't being clear on this.

The reason that we need to enforce a order in `mergeDims` is because the analysis here assumes the merge order. So virtual_innermost_dim need to enforce the same logical order, i.e. when projecting the virtual_innermost_dim to the rfactor domain on reference, we'll trim iterdomains off if they come in wrong order.

The reason for the insert at begin is that, `dims_to_merge` is in the opposite order that we merge to the innermost.
so we want the merged stuff to be.
`[leaf(dims_to_merge[i-1]), leaf(dims_to_merge[i-2]), ... leaf(dims_to_merge[0]), root(inner_most_dim])]`

^^^ yes, the index for `inner_most_dim` is retrieved on root domain instead of leaf domain, hence the inconsistency.

**Comment by jjsjann123** [comment]
!build

File: csrc/scheduler/vectorize_helper.cpp (hunk 12 above)
Comment (naoyam) [reply csrc/scheduler/vectorize_helper.cpp:876]: > i.e. when projecting the virtual_innermost_dim to the rfactor domain on reference

Are you referring to `projectId`? Didn't we talk about it just recently? Isn't the ordering of `ids` supposed not to matter?

Comment (jjsjann123) [reply csrc/scheduler/vectorize_helper.cpp:876]: Yes I was.

The order of `ids` does matter. The lifted restriction from the other PR is that, we don't want to restrict `ids` to be contiguous/vectorizable on the reference tensor.

I don't think I did a good job at explaining this in the original issue :stuck_out_tongue:  Nevertheless, here's the link to the original discussion: https://github.com/NVIDIA/Fuser/pull/753#discussion_r1301947487

File: csrc/scheduler/transpose.cpp (hunk 7 above)
Comment (naoyam) [reply csrc/scheduler/transpose.cpp:738]: Ah, OK, so, that means we can't figure out the vectorization factor for tensors that already have non-rfactor leaf domains, right? 

Can this be solved by projecting back from the rfactor to leaf domains?

Comment (jjsjann123) [reply csrc/scheduler/transpose.cpp:738]: I think so. It's just a refactor to have the propagator to use leaf instead of rfactor. I just don't know enough about how easy the new propagation is going to be, but would love to do this in next round. :office_worker:

**Review by naoyam — APPROVED** [review]
LGTM. Great work!
